src/ir/auth_logic/souffle_emitter.h

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2022 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef SRC_IR_AUTH_LOGIC_SOUFFLE_EMITTER_H_
+#define SRC_IR_AUTH_LOGIC_SOUFFLE_EMITTER_H_
+
+#include "absl/container/flat_hash_set.h"
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_join.h"
+#include "src/ir/auth_logic/datalog_ir.h"
+
+namespace raksha::ir::auth_logic {
+
+class SouffleEmitter {
+ public:
+  static std::string EmitProgram(const DLIRProgram& program) {
+    SouffleEmitter emitter;
+    std::string body = emitter.EmitProgramBody(program);
+    std::string outputs = emitter.EmitOutputs(program);
+    std::string decls = emitter.EmitDeclarations();
+    return absl::StrCat(std::move(body), "\n", std::move(outputs), "\n",
+                        std::move(decls), "\n");
+  }
+
+ private:
+  SouffleEmitter() = default;
+
+  // This function produces a normalized version of predicates to
+  // be used when generating declarations of the predicates. It replaces
+  // argument names with generic numbered ones. It is applied
+  // to predicates before they are added to the set of declarations so that
+  // there are no duplicate delcarations (which would otherwise happen
+  // whenever a predicate is referenced more than once with different
+  // arguments).
+  // To prevent instances of negated and non-negated uses of the predicate
+  // from generating two declarations, the sign here is always positive.
+  Predicate PredToDeclaration(const Predicate& predicate) {
+    std::vector<std::string> decl_args;
```

Comment:
It looks like you can use `MapIter`?

Reply:
ok I did that and also used it in one other place.